app/lib/report_helper.rb

```diff
@@ -1,23 +1,141 @@
   ZSCORE_MIN = -99
   ZSCORE_MAX =  99
   ZSCORE_WHEN_ABSENT_FROM_SAMPLE = -100
   ZSCORE_WHEN_ABSENT_FROM_BACKGROUND = 100
 
-  def external_report_info(report, view_level, params)
+  # TODO: For taxons that have no entry in the taxon_lineages table, we should
+  # substitute this value for genus_id, which allows us to group them
+  # all together;  this should match generate_aggregate_counts in
+  # app/models/pipeline_output.rb
+  MISSING_GENUS_ID = -200
+  MISSING_SPECIES_ID = -100
+  MISSING_SPECIES_ID_ALT = -1
+
+  # For taxon_count 'species' rows without a corresponding 'genus' rows,
+  # we create a fake singleton genus containing just that species;
+  # the fake genus IDs start here:
+  FAKE_GENUS_BASE = -1_900_000_000
+
+  DECIMALS = 1
+
+  DEFAULT_PARAMS = {
+    view_level:       'genus',
+    sort_by:          'highest_nt_aggregatescore',
+    threshold_zscore: 1.7,
+    threshold_rpm:    1.0,
+    threshold_r:      50,
+    threshold_percentidentity: 0.0,
+    threshold_alignmentlength: 0.0,
+    threshold_neglogevalue:    0.0,
+    threshold_aggregatescore:  0.0
+  }.freeze
+  IGNORED_PARAMS = [:controller, :action, :id].freeze
+
+  VIEW_LEVELS = %w[species genus].freeze
+  SORT_DIRECTIONS = %w[highest lowest].freeze
+  # We do not allow underscores in metric names, sorry!
+  METRICS = %w[r rpm zscore percentidentity alignmentlength neglogevalue aggregatescore].freeze
+  COUNT_TYPES = %w[NT NR].freeze
+  PROPERTIES_OF_TAXID = %w[tax_id name name_from_lineages name_from_counts tax_level genus_taxid category_name].freeze # note: no underscore in sortable column names
+  UNUSED_IN_UI_FIELDS = ['genus_taxid', :sort_key].freeze
+
+  # use the metric's NT <=> NR dual as a tertiary sort key (so, for example,
+  # when you sort by NT R, entries without NT R will be ordered amongst
+  # themselves based on their NR R (as opposed to arbitrary ordder);
+  # and within the Z, for things with equal Z, use the R as tertiary
+  OTHER_COUNT_TYPE = {
+    'NT' => 'NR',
+    'NR' => 'NT'
+  }.freeze
+  OTHER_METRIC = {
+    'zscore' => 'r',
+    'r' => 'zscore',
+    'rpm' => 'zscore'
+  }.freeze
+
+  def threshold_param?(param_key)
+    parts = param_key.to_s.split "_"
+    (parts.length == 2 && parts[0] == 'threshold' && METRICS.include?(parts[1]))
+  end
+
+  def decode_thresholds(params)
+    thresholds = {}
+    METRICS.each do |metric|
+      param_key = "threshold_#{metric}".to_sym
+      thresholds[metric] = params[param_key]
+    end
+    thresholds
+  end
+
+  def decode_sort_by(sort_by)
+    parts = sort_by.split "_"
+    return nil unless parts.length == 3
+    direction = parts[0]
+    return nil unless SORT_DIRECTIONS.include? direction
+    count_type = parts[1].upcase
+    return nil unless COUNT_TYPES.include? count_type
+    metric = parts[2]
+    return nil unless METRICS.include? metric
+    {
+      direction:    direction,
+      count_type:   count_type,
+      metric:       metric
+    }
+  end
+
+  def number_or_nil(string)
```

Comment:
what's this function for? can you just do string.to_f?

Reply:
string.to_f raises an exception when its argument is malformed;   i wanted it to return nil in that case

my bigger goal here was to simplify the JS and move argument validation & default values to rails